content/courses/ada-idioms/chapters/interrupt_handling.rst

```diff
@@ -0,0 +1,841 @@
+the `nul` character. Transmission will cease when the terminator character
+is encountered when sending a message object. Similarly, a message is
+considered complete when the terminator character is received. (The terminator
+is not stored in message content.)
+
+In a sense the interrupt handler is again communicating with tasks, but not
+directly, so entry parameters aren't applicable. Therefore, a
+:ada:`Suspension_Object` component is appropriate. But instead of one
+:ada:`Suspension_Object` variable, each :ada:`Message` object will contain two:
+one for notification of new content receipt, and one for notification of
+successful content transmission.
+
+For the sake of the Separation of Concerns principle, the type :ada:`Message`
+should be an ADT of its own, in a dedicated package:
+
+.. code-block:: ada
+
+    with Serial_IO;                     use Serial_IO;
+    with Ada.Synchronous_Task_Control;  use Ada.Synchronous_Task_Control;
+
+    package Message_Buffers is
+
+       type Message (Physical_Size : Positive) is tagged limited private;
+
+       function Content (This : Message) return String;
+       function Length (This : Message) return Natural;
+       procedure Set (This : in out Message;  To : String) with
+         Pre  => To'Length <= This.Physical_Size,
+         Post => Length (This) = To'Length and Content (This) = To;
+       ...
+       function Terminator (This : Message) return Character;
+       procedure Await_Transmission_Complete (This : in out Message);
+       procedure Await_Reception_Complete (This : in out Message);
+       procedure Signal_Transmission_Complete (This : in out Message);
+       procedure Signal_Reception_Complete (This : in out Message);
+
+    private
+
+       type Message (Physical_Size : Positive) is tagged limited record
+          Content               : String (1 .. Physical_Size);
+          Length                : Natural := 0;
+          Reception_Complete    : Suspension_Object;
+          Transmission_Complete : Suspension_Object;
+          Terminator            : Character := ASCII.NUL;
+       end record;
+
+    end Message_Buffers;
+
+
+In essence, a :ada:`Message` object is just the usual
+*variable length string* abstraction with a known terminator and ways to
+suspend and resume clients using them. Note the two
+:ada:`Suspension_Object` components.
+
+In this example the tasks to be notified are application tasks rather than
+second-level interrupt handlers. Client tasks can suspend themselves to await
+either transmission completion or reception completion. The :ada:`Message`
+procedures simply call the appropriate routines for the parameter's
+:ada:`Suspension_Object` components:
+
+.. code-block:: ada
+
+    procedure Await_Transmission_Complete (This : in out Message) is
+    begin
+       Suspend_Until_True (This.Transmission_Complete);
+    end Await_Transmission_Complete;
+    and likewise:
```

Comment:
"and likewise:" is meant to be regular text.

Reply:
Oops! That's a very good catch.